### PR TITLE
implements property values

### LINKED ISSUE #645: CRUD Property Values
API Design: https://github.com/helpwave/services/pull/619/files#diff-00a15fc20f35646f03a0864eb43abc5eaabf147aa91b5e434682d6266cffb72c and https://github.com/helpwave/services/pull/619/files#diff-294a770289827d8511107f61c3e1cef603a08c64762bb9ae0bfd4bb2ee8b08ec

Also see https://github.com/helpwave/services/issues/611

### PR DESCRIPTION
**Pull request by PaulKalho**
## Which issues does this pull request close?

closes #645

### CODE REVIEW

File: services/property-svc/repos/property_value_repo.sql [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        @subject_id,
+        sqlc.narg('text_value'),
+        sqlc.narg('number_value'),
+        sqlc.narg('bool_value'),
+        sqlc.narg('date_value'),
+        sqlc.narg('date_time_value'),
+        sqlc.narg('select_value')
+);
+
+-- name: GetPropertyValueBySubjectIDAndPropertyID :one
```
Comment (FoseFx) [inline services/property-svc/repos/property_value_repo.sql:16]: Unify this and GetPropertyValueByID and GetPropertyValuesBySubjectID into one query. If you want to you can only unify the latter two, as this query returns only the id, in that case rename this query to GetPropertyValueID.

File: services/property-svc/migrations/000002_property_values.up.sql [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+		) <= 1
+	),
+	UNIQUE (property_id, subject_id),
+	FOREIGN KEY (property_id)
+		REFERENCES properties(id)
+		ON DELETE CASCADE,
+
+	FOREIGN KEY (select_value)
+		REFERENCES select_options(id)
+		ON DELETE SET NULL
```
Comment (FoseFx) [inline services/property-svc/migrations/000002_property_values.up.sql:29]: maybe add a constraint that forces a select_option to belong to the property

File: proto/services/property_svc/v1/property_value_svc.proto [hunk 3]
```diff
@@ -2,55 +2,77 @@ syntax = "proto3";
...
-  string subject_type = 2;
-  string property_id = 3;
+message AttachPropertyValueRequest{
+  string subject_id = 1; // @gotags: validate:"uuid4"
+  string property_id = 2; // @gotags: validate:"uuid4"
+
   oneof value {
-    string text_value = 4; // FIELD_TYPE_TEXT
-    float number_value = 5; // FIELD_TYPE_NUMBER
+    string text_value = 4; // @gotags: validate:"required" // FIELD_TYPE_TEXT
```
Comment (FoseFx) [inline proto/services/property_svc/v1/property_value_svc.proto:18]: did you test this? or does the "required" also force a value to exist

File: proto/services/property_svc/v1/property_value_svc.proto [hunk 4]
```diff
@@ -2,55 +2,77 @@ syntax = "proto3";
...
+    optional Set set = 7;
+    */
+
+    oneof value {
+      string text_value = 9; // FIELD_TYPE_TEXT
+      double number_value = 10; // FIELD_TYPE_NUMBER
+      bool bool_value = 11; // FIELD_TYPE_CHECKBOX
+      Date date_value = 12; // FIELD_TYPE_DATE
+      google.protobuf.Timestamp date_time_value = 13; // FIELD_TYPE_DATE_TIME
+      string select_value = 14; // FIELD_TYPE_SELECT, id of option
```
Comment (FoseFx) [inline proto/services/property_svc/v1/property_value_svc.proto:62]: move those ids down to 6+

File: libs/hwutil/assert.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	"time"
+)
+
+// AssertTimestampToTime takes an interface and returns a time
+func AssertTimestampToTime(i interface{}) (*time.Time, error) {
+	var s float64
+	var n float64
+	d, ok := i.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("could not assert map")
```
Comment (FoseFx) [inline libs/hwutil/assert.go:14]: return localized errors here (see task-svc for examples)

File: services/property-svc/internal/property-value/api/grpc.go [hunk 6]
```diff
@@ -18,71 +23,105 @@ func NewPropertyValueService(aggregateStore hwes.AggregateStore) *PropertyValueG
...
 		return nil, err
 	}
 
-	response := &pb.GetPropertyValueResponse{
-		Id:          property.ID.String(),
-		SubjectType: property.SubjectType,
-		PropertyId:  property.PropertyID.String(),
-	}
+	propertyValuesRes, err := hwutil.MapWithErr(propertyValues, func(propertyValue property_value_repo.PropertyValue) (*pb.GetAttachedPropertyValuesResponse_Value, error) {
+		property, err := propertyRepo.GetPropertyById(ctx, propertyValue.PropertyID)
```
Comment (FoseFx) [inline services/property-svc/internal/property-value/api/grpc.go:82]: database query in for loop, big no-no! refactor into building *one* query

maybe you can even do this in a JOIN together with the previous query

File: services/property-svc/internal/property-value/commands/v1/attach_property_value.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+)
+
+type AttachPropertyValueCommandHandler func(ctx context.Context, propertyValueID uuid.UUID, propertyID uuid.UUID, value interface{}, subjectID uuid.UUID) error
+
+func NewAttachPropertyValueCommandHandler(as hwes.AggregateStore) AttachPropertyValueCommandHandler {
+	return func(ctx context.Context, propertyValueID uuid.UUID, propertyID uuid.UUID, value interface{}, subjectID uuid.UUID) error {
+		propertyValueRepo := property_value_repo.New(hwdb.GetDB())
+		var a *aggregate.PropertyValueAggregate
+
+		// TODO: ensure eventual consistency by not using the projection here
```
Comment (FoseFx) [inline services/property-svc/internal/property-value/commands/v1/attach_property_value.go:19]: Eventual consistency can not be "ensured", we already have it

File: services/property-svc/internal/property-value/commands/v1/attach_property_value.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+		})
+		if err := hwdb.Error(ctx, err); err != nil {
+			return err
+		}
+
+		if existingPropertyValueID != nil {
+			if a, err = aggregate.LoadPropertyValueAggregate(ctx, as, *existingPropertyValueID); err != nil {
+				return err
+			}
+			// TBD: update value will be triggered, even if the value is not the type the property defines
```
Comment (FoseFx) [inline services/property-svc/internal/property-value/commands/v1/attach_property_value.go:32]:
```suggestion
			// TODO: update value will be triggered, even if the value is not the type the property defines
```

**Review by FoseFx — CHANGES_REQUESTED** [review]

File: proto/services/property_svc/v1/property_value_svc.proto (hunk 3 above)
Comment (PaulKalho) [reply proto/services/property_svc/v1/property_value_svc.proto:18]: Yes, if the string is empty we will get an validation-error

Comment (FoseFx) [reply proto/services/property_svc/v1/property_value_svc.proto:18]: I sure hope so, but how does it behave when no string is provided at all?

Comment (PaulKalho) [reply proto/services/property_svc/v1/property_value_svc.proto:18]: It allows no string as well.

File: services/property-svc/internal/property-value/commands/v1/attach_property_value.go (hunk 7 above)
Comment (PaulKalho) [reply services/property-svc/internal/property-value/commands/v1/attach_property_value.go:19]: right

File: services/property-svc/repos/property_value_repo.sql (hunk 1 above)
Comment (PaulKalho) [reply services/property-svc/repos/property_value_repo.sql:16]: `GetPropertyValuesBySubjectID` now also embeds `property` now, not only the `propertyValue`. I consider this as resolved.

File: services/property-svc/migrations/000002_property_values.up.sql [hunk 9]
```diff
@@ -0,0 +1,52 @@
+CREATE OR REPLACE FUNCTION validate_select_option(property_id uuid, select_value uuid)
```
[the remark this answers is not in the document]
Comment (FoseFx) [reply services/property-svc/migrations/000002_property_values.up.sql:1]: e.g. `does_select_option_belong_to_property`

File: services/property-svc/migrations/000002_property_values.up.sql [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+	ELSE
+		RETURN TRUE;
+	END IF;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TABLE IF NOT EXISTS property_values (
+	id uuid PRIMARY KEY DEFAULT uuid_generate_v4(),
+	property_id uuid NOT NULL,
+	subject_id uuid NOT NULL,
```
Comment (FoseFx) [inline services/property-svc/migrations/000002_property_values.up.sql:21]: we frequently do lookups by `(property_id, subjectid)` so we should add an [Index](https://www.postgresql.org/docs/current/sql-createindex.html)

File: proto/services/property_svc/v1/property_value_svc.proto [hunk 11]
```diff
@@ -2,55 +2,77 @@ syntax = "proto3";
...
-    bool bool_value = 6; // FIELD_TYPE_CHECKBOX
-    string date_value = 7; // FIELD_TYPE_DATE TODO: is there a proto date type?
-    string date_time_value = 8; // FIELD_TYPE_DATE_TIME
-    string select_value = 9; // FIELD_TYPE_SELECT, id of option
+    string text_value = 3; // @gotags: validate:"required" // FIELD_TYPE_TEXT
+    double number_value = 4; // FIELD_TYPE_NUMBER
+    bool bool_value = 5; // FIELD_TYPE_CHECKBOX
+    Date date_value = 6; // FIELD_TYPE_DATE
+    google.protobuf.Timestamp date_time_value = 7; // FIELD_TYPE_DATE_TIME
+    string select_value = 10; // FIELD_TYPE_SELECT, id of option
```
Comment (FoseFx) [inline proto/services/property_svc/v1/property_value_svc.proto:23]:
```suggestion
    string select_value = 8; // FIELD_TYPE_SELECT, id of option
```

File: services/property-svc/internal/property-value/queries/v1/get_property_values_by_subject_id.go [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+	"property-svc/internal/property-value/models"
+	"property-svc/repos/property_value_repo"
+)
+
+type GetPropertyValuesWithPropertiesBySubjectIQueryHandler func(ctx context.Context, subjectID uuid.UUID) ([]models.PropertyValueWithProperty, error)
+
+func NewGetPropertyValuesWithPropertiesBySubjectIDQueryHandler(propertyValueRepo *property_value_repo.Queries) GetPropertyValuesWithPropertiesBySubjectIQueryHandler {
+	return func(ctx context.Context, subjectID uuid.UUID) ([]models.PropertyValueWithProperty, error) {
+
+		propertyValuesWithProperties, err := hwdb.Optional(propertyValueRepo.GetPropertyValuesWithPropertyBySubjectID)(ctx, subjectID)
```
Comment (FoseFx) [inline services/property-svc/internal/property-value/queries/v1/get_property_values_by_subject_id.go:18]: [sqlc.yaml](https://github.com/helpwave/services/blob/main/services/property-svc/sqlc.yaml#L13) Optional is not needed

File: services/property-svc/repos/property_value_repo.sql [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+	select_value = sqlc.narg('select_value')
+WHERE id = $1;
+
+-- name: GetPropertyValueByID :one
+SELECT * FROM property_values WHERE id = $1;
+
+-- name: GetPropertyValuesWithPropertyBySubjectID :many
+SELECT
+	sqlc.embed(properties),
+	sqlc.embed(property_values)
```
Comment (FoseFx) [inline services/property-svc/repos/property_value_repo.sql:37]: LEFT JOIN embeds dont work. property_values may not contain any items that match https://togithub.com/sqlc-dev/sqlc/issues/2348

**Review by FoseFx — CHANGES_REQUESTED** [review]

File: services/property-svc/migrations/000002_property_values.up.sql [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+	-- check if at most one value field is set
+	check(
+		(
+			(text_value is not null)::integer +
+			(number_value is not null)::integer +
+			(bool_value is not null)::integer +
+			(date_value is not null)::integer +
+			(date_time_value is not null)::integer +
+			(select_value is not null)::integer
+		) <= 1
```
[the remark this answers is not in the document]
Comment (PaulKalho) [reply services/property-svc/migrations/000002_property_values.up.sql:37]: i see.

**Review by FoseFx — APPROVED** [review]

**Comment by FoseFx** [comment]
🎉

**Pull request merged**